### PR TITLE
Homework for Lesson 7

### PR DESCRIPTION
**Pull request by filkodaria**
Homework_7: task_1, task_2, task_3.

### CODE REVIEW

File: hw_7/hw_7.1.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+6. Проверить с массивом const arr = [NaN, 1, true, 5, "hello", undefined, 15.5, {}, []];
+*/
+
+function getOnlyNumbers(arrayOnAnyValues) {
+	if (!(Array.isArray(arrayOnAnyValues))) {
+		throw new Error(`Something went wrong! Only array format for entered data is valid.`);
+	} else {
+		const result = arrayOnAnyValues.filter(el => typeof el === 'number' && !isNaN(el));
+		return result; 
+	}
```
Comment (VladimirLQA) [inline hw_7/hw_7.1.js]:
```suggestion
	}
     const result = arrayOnAnyValues.filter(el => typeof el === 'number' && !isNaN(el));
      return result; 
```

File: hw_7/hw_7.2.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+const str = 'I am the best AQA ever!';
+
+function transformToAppearanceTimes(string) {
+	if (typeof string !== 'string') {
+		throw new Error('Something went wrong! Only string type for entered data is valid.');
+	} else {
+		const arr = str.toLowerCase().split('');
+
+		const transformedArray = arr.map((el, ind, arr) => {
+			if (el.charCodeAt(0) >= 97 && el.charCodeAt(0) <= 122) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]:
```js
  if(el.match(/[a-z]/g))
````

File: hw_7/hw_7.2.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+				let counter = 0;
+
+				for (let i = 0; i < arr.length; i++) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]: Can use `.forEach` method for comparing chars

File: hw_7/hw_7.2.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+const prices = [64, 7556, 345, 7556, 345, 7556, 345, 7556, 433, 345, 756, 123, 942, 3112, 421, 9341, 1212, 8, 43, 41, 345, 341, 21, 321, 123];
+
+function getPriceInfo(priceTotal, priceAverage) {
+	return `Total price of all items is ${priceTotal}$. Average price of item is ${priceAverage}$.`;
+}
+
+function getTotalPrice(arr) {
+	const sum = arr.reduce((total, value) => total + value, 0);
+	return sum;
+}
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]: Try to use `arrow` functions :)
```suggestion
const getTotalPrice = (arr) => arr.reduce((total, value) => total + value, 0);
```

File: hw_7/hw_7.2.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+// sabtask 3
+/*
+3. Напишите функцию, которая принимает на вход массив слов и возвращает отсортированный массив по следующему критерию: количество гласных букв. 
+Массив должен быть отсортирован по возврастанию количества гласных букв в слове.
+*/
+
+const words = ['learn', 'Javascript', 'no', 'sleep', 'night'];
+
+function sortArrayByVowels(arr) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]: You can make that helper function will accept the string and return the count of vowels only
For example
```js
const countVowels = (str) => count;
const sortArrayByVowels = (arr) => [...arr].sort((a, b) => countVowels(a) - countVowels(b));
```

File: hw_7/hw_7.2.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	return sortedArray.map(el => el[1]);
+}
+
+function findVowelsAmount(arr) {
+	const vowels = 'aeiouy';
+	
+	const amountVowelsArray = arr.map(el => {
+		let counterForVowels = 0;
+
+		for (let i = 0; i < el.length; i++) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]: Again we can use efficient array methods :)

File: hw_7/hw_7.2.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  Усложнение: в массиве могут быть вложены еще массивы на разной глубине. (ПОПРОБУЙТЕ МЕТОД .flat(), изучите как он работает с разными глубинами вложенности)
+  Вернуть нужно всё также есть ли у каждой скобочки соответствующая пара. 
+  Пример:
+  const arr = [[['(']], ')', '(', ')', ')', ['(', ['('], [')']]]
+*/
+
+const brackets = [ '(', ')', '(', '(', '(', ')', ')', ')'];
+const brackets2 = [ '(', '(', '(', '(', ')', ')', ')'];
+
+function checkBracketsPair(arr) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js:92]: Follow the task description
Try this example
```
const arr = [[['(']], ')', '(', ')', ')', ['(', ['('], [')']]]
```

File: hw_7/hw_7.2.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	let newString = arr.join(''); 
+
+  while (newString.includes('()')) {
+    newString = newString.replace('()', '');
+  }
+
+	if (!newString.length) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.2.js]: Ternary operator

File: hw_7/hw_7.3.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+// subtask 5
+/*
+5*. Напишите функцию findMissingNumber(arr), которая принимает массив чисел от 1 до N (исключая одно число) 
+  и возвращает пропущенное число. Массив не отсортирован и может содержать дубликаты. 
+  Решите эту задачу, используя эффективные методы массива.
+*/
+
+const numbers = [1, 4, 7, 9, 3, 4, 9, 10, 11, 2, 6, 6, 8, 12]  // 5 is excluded
+
+function findMissingNumber(arr) {
```
Comment (VladimirLQA) [inline hw_7/hw_7.3.js:12]: Try to use efficient array methods :)

**Review by VladimirLQA — CHANGES_REQUESTED** [review]

**Comment by filkodaria** [comment]
@VladimirLQA 
please, check changes to tasks 7.1 and 7.2.

File: hw_7/hw_7.2.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        if (item === el) {
+          counter++ ;
+        }
+      })
+			return counter;
+		} else return el;
+	})	
+
+	return transformedArray.join('');
+}
```
Comment (anatoly-karpovich) [inline hw_7/hw_7.2.js:30]:
```suggestion
function transformToAppearanceTimes(string) {
  if (typeof string !== "string") throw new Error("Argument must be a string");
  const abc = "abcdefghijklmnopqrstuvwxyz";
  string = string.toLowerCase()
  return [...string].map((char, i, arr) => (abc.includes(char) ? arr.filter((c) => c === char).length : char)).join("");
}
```

File: hw_7/hw_7.3.js (hunk 9 above)
Comment (anatoly-karpovich) [reply hw_7/hw_7.3.js:12]: Agree with magister VoYoda)

**Review by anatoly-karpovich — APPROVED** [review]

**Review by VladimirLQA — APPROVED** [review]

**Comment by filkodaria** [comment]
@VladimirLQA 
I've added changes to task 7.3 (subtask 5)
Check please

**Review by VladimirLQA — APPROVED** [review]

**Comment by filkodaria** [comment]
Thank you guys for homework review!

**Pull request merged**